src/Mvc/src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -119,6 +123,21 @@ private void EnsureServer()
 
             EnsureDepsFile();
 
+            var hostBuilder = HostFactoryResolver.ResolveHostBuilderFactory<IHostBuilder>(typeof(TEntryPoint).Assembly)?.Invoke(Array.Empty<string>());
+            if (hostBuilder != null)
+            {
+                hostBuilder.ConfigureWebHost(webHostBuilder =>
+                {
+                    webHostBuilder.UseEnvironment("Development");
+                    SetContentRoot(webHostBuilder);
+                    _configuration(webHostBuilder);
+                    webHostBuilder.UseTestServer();
```

Comment:
What happens if we call webHostBuilder.UseKestrel() after this. Do all services get replaced and everything works just fine? If that's the case, I'd be more inclined to put UseTestServer above _configuration and to change _server (and the associated property) to be an IServer instance. 

Note this would be a breaking change, but most people using this class don't use TestServer directly. It's there just as a scape hatch in case you need to create something like a websocket or similar.

Reply:
TestServer is a lot more useful than IServer. Calling UseKestrel after UseTestServer replace the server and prevent CreateClient from working.